addons/folo/common/src/main/java/org/commonjava/indy/folo/data/FoloRecordCassandra.java

```diff
@@ -0,0 +1,426 @@
+
+        logger.warn("-- Cassandra Folo Records Keyspace and Tables created  ---");
+
+    }
+
+    @Override
+    public boolean recordArtifact(TrackedContentEntry entry) throws FoloContentException, IndyWorkflowException {
+
+        String buildId = entry.getTrackingKey().getId();
+        String path = entry.getPath();
+
+//        String api = uriInfo.getBaseUriBuilder().path("api").build().toString();
+        String content  = "content";
+        String packageType = entry.getStoreKey().getPackageType();
+        String storeKey = entry.getStoreKey().getType().singularEndpointName();
+        String name = entry.getStoreKey().getName();
+
+
+        try {
+            String localUrl;
+
+            localUrl = UrlUtils.buildUrl("" ,content , packageType , storeKey ,name , path );
+
+            BoundStatement bind = getTrackingRecordByBuildIdAndPath.bind(buildId,path);
+            ResultSet trackingRecord = session.execute(bind);
+            Row one = trackingRecord.one();
+
+            if(one!=null) {
+                DtxTrackingRecord dtxTrackingRecord = one.get(0, DtxTrackingRecord.class);
+                Boolean state = dtxTrackingRecord.getState();
+
+                if(state) {
+                    throw new FoloContentException( "Tracking record: {} is already sealed!", entry.getTrackingKey() );
+                }  else {
+
+                    DtxTrackingRecord dtxTrackingRecord1 =
+                            DtxTrackingRecord.fromTrackedContentEntry(entry,false);
+
+                    if(dtxTrackingRecord.getTrackingKey().equals(dtxTrackingRecord1.getTrackingKey()) &&
+                        dtxTrackingRecord.getPath().equals(dtxTrackingRecord1.getPath())) {
+
+                        dtxTrackingRecord1.setLocalUrl(localUrl);
+
+                        trackingMapper.save(dtxTrackingRecord1);
+                        return true;
+                    } else {
+                        return false;
+                    }
+                }
+
+            } else {
+                DtxTrackingRecord dtxTrackingRecord = new DtxTrackingRecord(entry);
+                dtxTrackingRecord.setLocalUrl(localUrl);
+                trackingMapper.save(dtxTrackingRecord); //  optional Options with TTL, timestamp...
+                return true;
+            }
+
+        } catch (MalformedURLException e) {
+            throw new IndyWorkflowException( "Mailformed URL  for record: {}  path: {}", entry.getTrackingKey(),entry.getPath() );
+        }
+
+
+    }
+
+    @Override
+    public void delete(TrackingKey key) {
+
+        // Do we  want to remove records from DB?
+
+    }
+
+    @Override
+    public void replaceTrackingRecord(TrackedContent record) {
+
+        saveTrackedContentRecords(record);
+
+    }
+
+    @Override
+    public boolean hasRecord(TrackingKey key) {
+        return hasSealedRecord(key) || hasInProgressRecord(key);
```

Comment:
This seems like a pretty inefficient way to deal with this in Cassandra. Can we simply select a maximum of 1 row where the tracking key matches?

Reply:
> We should probably have storekey and possibly store_effect as cluster keys (in addition to path) though, right? So we can filter more efficiently on those when we query, without having to resort to ALLOW FILTERING?

If we collapse the sealed records table into the path records table, then I guess there would be an index on sealed in that table?

 I've answered this. little up on this page.

> This seems like a pretty inefficient way to deal with this in Cassandra. Can we simply select a maximum of 1 row where the tracking key matches?

It's the. same. thing because of optimal table usage. of partition and. primary keys:
All records from one tracking_key will be. stored on one node from cassandra cluster so it's. no. matter if you are getting one record or all of them ( from that tracking_key ) query performance will. be negligible.
BTW if we are going from that logic then we don't. need to. search in both (sealed and. inProgress) records , it will be enough if we just query table for that key and just. get one record and return true if exist or false if not.